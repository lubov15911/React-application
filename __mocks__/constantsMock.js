export const moviePreviewData = {
    poster_path: 'https://image.tmdb.org/t/p/w500/eKi8dIrr8voobbaGzDpe8w0PVbC.jpg',
    title: 'Coco',
    vote_average: 7.8,
    tagline: 'The celebration of a lifetime',
    release_date: '2017-10-27',
    runtime: 105,
    genres: ["Adventure", "Comedy", "Family", "Animation"],
    overview: 'Despite his familyâ€™s baffling generations-old ban on music, Miguel dreams of becoming an accomplished musician like his idol, Ernesto de la Cruz.',
};

export const filmsData = [{
    id: 354912,
    poster_path: 'https://image.tmdb.org/t/p/w500/eKi8dIrr8voobbaGzDpe8w0PVbC.jpg',
    title: 'Coco',
    release_date: '2017-10-27',
    genres: ["Adventure", "Comedy", "Family", "Animation"],
}, {
    id: 269149,
    poster_path: "https://image.tmdb.org/t/p/w500/sM33SANp9z6rXW8Itn7NnG1GOEs.jpg",
    title: "Zootopia",
    release_date: "2016-02-11",
    genres: ["Animation", "Adventure", "Family", "Comedy"],
}, {
    id: 333339,
    poster_path: 'https://image.tmdb.org/t/p/w500/pU1ULUq8D3iRxl1fdX2lZIzdHuI.jpg',
    title: 'Ready Player One',
    release_date: '2018-03-28',
    genres: ["Adventure", "Science Fiction", "Action"],
}, {
    id: 337167,
    poster_path: "https://image.tmdb.org/t/p/w500/3kcEGnYBHDeqmdYf8ZRbKdfmlUy.jpg",
    title: "Fifty Shades Freed",
    release_date: "2018-03-28",
    genres: ["Drama", "Romance"],
}];

export const searchOptionsData = {
    title: 'Title',
    genre: 'Genre'
};

export const sortOptionsData = {
    rating: 'Rating',
    release: 'Release date'
};

export const searchValueData = 'Ready';
