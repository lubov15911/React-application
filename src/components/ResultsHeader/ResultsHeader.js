import React from 'react';
import PropTypes from 'prop-types';

import ToggleComponent from '../ToggleComponent/ToggleComponent';

import './ResultsHeader.scss';

const SortOptions = {
    first: 'Release date',
    second: 'Rating'
};

export default class ResultsHeader extends React.Component {
    constructor(props) {
        super(props);
        this.state = {
            sortOption: SortOptions.first
        };
        this.toggleSort = this.toggleSort.bind(this);
    }

    toggleSort() {
        const { sortOption } = this.state;
        const { handleSortOption } = this.props;
        const newValue = (sortOption === SortOptions.first) ? SortOptions.second : SortOptions.first;
        this.setState({ sortOption: newValue });
        handleSortOption(newValue);
    }

    render() {
        const { totalResults } = this.props;
        const { sortOption } = this.state;

        return (
            <div className="results-header">
                {totalResults && <p><b>{totalResults} movie found</b></p>}
                <ToggleComponent
                    className="toggle1"
                    toggleType='Sort'
                    options={SortOptions}
                    selected={sortOption}
                    handleToggle={this.toggleSort} />
            </div>
        );
    }
}
ResultsHeader.propTypes = {
    totalResults: PropTypes.number.isRequired,
    handleSortOption: PropTypes.func.isRequired,
};
