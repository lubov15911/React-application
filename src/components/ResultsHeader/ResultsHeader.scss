@import "../../_colors";

.results-header {
  background: $gray;
  padding: 0 12em;
  height: 7em;
  display: flex;
  justify-content: space-between;
  align-items: center;

  & > p {
    font-size: 1.5em;
  }

  .toggler {
    margin-left: auto;
  }
}
