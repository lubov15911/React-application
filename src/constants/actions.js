export const SAVE_FILM_LIST = 'SAVE_FILM_LIST';
export const OPEN_MOVIE_PAGE = 'OPEN_MOVIE_PAGE';
export const UPDATE_SORT_OPTION = 'UPDATE_SORT_OPTION';
export const UPDATE_SEARCH_OPTION = 'UPDATE_SEARCH_OPTION';
export const UPDATE_SEARCH_VALUE = 'UPDATE_SEARCH_VALUE';
