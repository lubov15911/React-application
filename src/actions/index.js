import { call, put, all, takeLatest } from 'redux-saga/effects';

import { URL } from '../constants';
import * as Constants from '../constants/actions';

const fetch = require('node-fetch');
const encodeurl = require('encodeurl');

// Internal functions
function getURL({ searchValue, searchOption, sortOption }) {
    let url = `${URL}movies`;
    const queries = [];
    if (sortOption) {
        queries.push(`sortBy=${sortOption.toLowerCase().replace(' ', '_')}`);
    }
    if (searchValue) {
        queries.push(`search=${searchValue}`);
    }
    if (searchOption) {
        queries.push(`searchBy=${searchOption.toLowerCase()}`);
    }

    if (queries.length) {
        url += `?${queries.join('&')}`;
    }

    return encodeurl(url);
}

// End Internal functions

// films actions
const saveFilmList = (list) => ({
    type: Constants.SAVE_FILM_LIST,
    payload: list,
});

export const getMovies = (value) => ({
    type: Constants.GET_MOVIES,
    payload: value,
});

// Saga
export function* asyncGetMovies(action) {
    const response = yield call(fetch, getURL(action.payload));
    const data = yield response.json();

    yield put(saveFilmList(data.data));
}

export function* watchGetMovies() {
    yield takeLatest(Constants.GET_MOVIES, asyncGetMovies);
}

// movieData actions
const openMoviePage = (movieData) => ({
    type: Constants.OPEN_MOVIE_PAGE,
    payload: movieData,
});

export const getMovieData = (value) => ({
    type: Constants.GET_MOVIE_DATA,
    payload: value,
});

// Saga
export function* asyncGetMovieData(action) {
    const response = yield call(fetch, `${URL}movies/${action.payload}`);
    const data = yield response.json();

    yield put(openMoviePage(data));
}

export function* watchGetMovieData() {
    yield takeLatest(Constants.GET_MOVIE_DATA, asyncGetMovieData);
}

export function* appSaga() {
    yield all([
        watchGetMovies(),
        watchGetMovieData(),
    ]);
}

// searchValue actions
export const updateSearchValue = (value) => ({
    type: Constants.UPDATE_SEARCH_VALUE,
    payload: value,
});

// searchOption/sortOption actions
export const updateSortOption = (value) => ({
    type: Constants.UPDATE_SORT_OPTION,
    payload: value,
});

export const updateSearchOption = (value) => ({
    type: Constants.UPDATE_SEARCH_OPTION,
    payload: value,
});

export const resetState = () => ({
    type: Constants.RESET_PAGE_STATE
});
