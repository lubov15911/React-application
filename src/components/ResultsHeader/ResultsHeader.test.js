import React from 'react';
import configureStore from 'redux-mock-store';
import { shallow } from 'enzyme';

import ResultsHeader from './ResultsHeader';

import { updateSortOption } from '../../actions';
import { filmsData, sortOptionsData, moviePreviewData } from '../../../__mocks__/constantsMock';

describe('ResultsHeader', () => {
    let mockStore, initState, component;

    beforeEach(() => {
        initState = {
            films: filmsData,
            sortOption: sortOptionsData.release,
            movieData: moviePreviewData,
        };
    });

    it('should render correctly without movieData', () => {
        initState.movieData = null;
        mockStore = configureStore()(initState);
        component = shallow(<ResultsHeader store={mockStore} />).dive();

        expect(component).toMatchSnapshot();
    });

    it('should render correctly with movieData', () => {
        mockStore = configureStore()(initState);
        component = shallow(<ResultsHeader store={mockStore} />).dive();

        expect(component).toMatchSnapshot();
    });

    it('should handle toggle option', () => {
        const toggleEvent = { currentTarget: { value: sortOptionsData.rating } };

        mockStore = configureStore()(initState);
        mockStore.dispatch = jest.fn();

        component = shallow(<ResultsHeader store={mockStore} />).dive();
        component.find('ToggleComponent').props().handleToggle(toggleEvent);

        expect(mockStore.dispatch).toHaveBeenCalledWith(updateSortOption(sortOptionsData.rating));

        jest.clearAllMocks();
    });
});
