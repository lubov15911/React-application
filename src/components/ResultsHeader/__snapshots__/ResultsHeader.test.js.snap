// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ResultsHeader should render correctly with movieData 1`] = `
<div
  className="results-header"
>
  <p>
    Films by 
    Adventure
     genre
  </p>
  <ToggleComponent
    className="toggle1"
    handleToggle={[Function]}
    options={
      Object {
        "first": "Release date",
        "second": "Rating",
      }
    }
    selected="Release date"
    toggleType="Sort "
  />
</div>
`;

exports[`ResultsHeader should render correctly without movieData 1`] = `
<div
  className="results-header"
>
  <p>
    <b>
      4
       movie found
    </b>
  </p>
  <ToggleComponent
    className="toggle1"
    handleToggle={[Function]}
    options={
      Object {
        "first": "Release date",
        "second": "Rating",
      }
    }
    selected="Release date"
    toggleType="Sort "
  />
</div>
`;
